src/components/implementation/no_interface/scheddev/sl.c

```diff
@@ -61,40 +83,103 @@ sl_thd_block(thdid_t tid)
 
 	sl_cs_enter();
 	t = sl_thd_curr();
-	if (unlikely(t->state == SL_THD_WOKEN)) {
-		t->state = SL_THD_RUNNABLE;
+	if (sl_thd_block_no_cs(t, 0)) {
 		sl_cs_exit();
 		return;
 	}
-
-	assert(t->state == SL_THD_RUNNABLE);
-	t->state = SL_THD_BLOCKED;
-	sl_mod_block(sl_mod_thd_policy_get(t));
 	sl_cs_exit_schedule();
 
 	return;
 }
 
-void
-sl_thd_wakeup(thdid_t tid)
+cycles_t
+sl_thd_block_timeout(thdid_t tid, cycles_t abs_timeout)
 {
+	cycles_t jitter = 0;
 	struct sl_thd *t;
-	thdcap_t       thdcap;
-	tcap_t         tcap;
-	tcap_prio_t    prio;
+	int is_timeout = abs_timeout ? 1 : 0;
+
+	/* TODO: dependencies not yet supported */
+	assert(!tid);
 
 	sl_cs_enter();
-	t = sl_thd_lkup(tid);
-	if (unlikely(!t)) goto done;
+	t = sl_thd_curr();
+	if (sl_thd_block_no_cs(t, is_timeout)) {
+		sl_cs_exit();
+		return 0;
+	}
+	if (is_timeout) sl_timeout_block(t, abs_timeout);
+	sl_cs_exit_schedule();
+
+	/* FIXME: possible race around these accesses */
+	if (t->wakeup_cycs > t->timeout_cycs) jitter = t->wakeup_cycs - t->timeout_cycs;
+
+	return jitter;
+}
+
+unsigned
+sl_thd_block_periodic(thdid_t tid)
+{
+	unsigned jitter = 0;
```

Comment:
Kk, will keep that in mind for style.

`cycles_t` as return type could be misleading and a misnomer as I'm returning `number of periods` elapsed in through this API, is also one of the reasons I introduced a separate API `sl_thd_block_periodic()`. If their return types can be same, I could use `sl_thd_block_timeout()` with `timeout == 0` to be periodic, which was my originally how I thought I could do it. Because I don't somehow like that I could block indefinitely in two different ways with the current API, one through `sl_thd_block()` and other through `sl_thd_block_timeout(...,0)`.